packages/assets-controllers/src/TokenBalancesController.test.ts

```diff
@@ -1,191 +1,250 @@
+
     await tokenBalances.updateBalances();
-    expect(mock.called).toBe(false);
+
+    expect(tokenBalances.state.contractBalances).toStrictEqual({
+      '0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0': '1',
+    });
+
+    tokenBalances.disable();
+    triggerTokensStateChange({
+      ...getDefaultTokensState(),
+      tokens: [
+        {
+          address: '0x00',
+          symbol: 'FOO',
+          decimals: 18,
+        },
+      ],
+    });
+
+    expect(tokenBalances.state.contractBalances).toStrictEqual({
+      '0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0': '1',
+    });
   });
 
   it('should clear previous interval', async () => {
-    const mock = sinon.stub(global, 'clearTimeout');
-    const tokenBalances = new TokenBalancesController(
-      {
-        onTokensStateChange: sinon.stub(),
-        getSelectedAddress: () => '0x1234',
-        getERC20BalanceOf: sinon.stub(),
-      },
-      { interval: 1337 },
-    );
-    await tokenBalances.poll(1338);
-    await advanceTime({ clock, duration: 1339 });
-    expect(mock.called).toBe(true);
+    const controller = new TokenBalancesController({
+      interval: 1337,
+      onTokensStateChange: jest.fn(),
+      getSelectedAddress: () => '0x1234',
+      getERC20BalanceOf: jest.fn(),
+      messenger: getMessenger(),
+    });
+
+    const mockClearTimeout = jest.spyOn(global, 'clearTimeout');
+
+    await controller.poll(1338);
+
+    jest.advanceTimersByTime(1339);
+
+    expect(mockClearTimeout).toHaveBeenCalled();
   });
 
   it('should update all balances', async () => {
     const selectedAddress = '0x0000000000000000000000000000000000000001';
     const address = '0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0';
-    const tokenBalances = new TokenBalancesController(
-      {
-        onTokensStateChange: jest.fn(),
-        getSelectedAddress: () => selectedAddress,
-        getERC20BalanceOf: sinon.stub().returns(new BN(1)),
-      },
-      {
-        interval: 1337,
-        tokens: [{ address, decimals: 18, symbol: 'EOS', aggregators: [] }],
-      },
-    );
-    expect(tokenBalances.state.contractBalances).toStrictEqual({});
+    const controller = new TokenBalancesController({
+      interval: 1337,
+      tokens: [{ address, decimals: 18, symbol: 'EOS', aggregators: [] }],
+      onTokensStateChange: jest.fn(),
+      getSelectedAddress: () => selectedAddress,
+      getERC20BalanceOf: jest.fn().mockReturnValue(new BN(1)),
+      messenger: getMessenger(),
+    });
 
-    await tokenBalances.updateBalances();
-    const mytoken = getToken(tokenBalances, address);
-    expect(mytoken?.balanceError).toBeNull();
-    expect(Object.keys(tokenBalances.state.contractBalances)).toContain(
-      address,
-    );
+    expect(controller.state.contractBalances).toStrictEqual({});
+
+    await controller.updateBalances();
 
-    expect(
-      tokenBalances.state.contractBalances[address].toNumber(),
-    ).toBeGreaterThan(0);
+    const mytoken = getToken(controller, address);
```

Comment:
Ah, I see this is the reason why you had to add a `getTokens` method to the controller. The behavior in `updateBalances` seems very strange to me — it seems to be directly modifying tokens in memory. Whatever the reason, you can take advantage of this without having to add a new method to the controller: if you extract the token in `[{ address, decimals: 18, symbol: 'EOS', aggregators: [] }]` to a new variable, then you can test that it gets changed directly in this test. As a simple example:

``` typescript
const token = {
  address: '0x86fa049857e0209aa7d9e616f7eb3b3b78ecfdb0',
  decimals: 18,
  symbol: 'EOS',
  aggregators: [],
};
const controller = new TokenBalancesController({
  interval: 1337,
  tokens: [token],
  onTokensStateChange: jest.fn(),
  getSelectedAddress: () => selectedAddress,
  getERC20BalanceOf: jest.fn().mockReturnValue(new BN(1)),
  messenger: getMessenger(),
});

await controller.updateBalances();

// `token` magically gets updated
expect(token.balanceError).toBeNull();
```